### PR TITLE
Test development guidelines

### PR DESCRIPTION
**Pull request by eder7**
See [rich view](https://github.com/mysugr/sweetest/blob/guidelines/GUIDELINES.md)

### CODE REVIEW

File: GUIDELINES.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+After some time working with sweetest we came up to the conclusion that it leaves a lot of freedom to design tests as we want. But on the other side there is an increasing need for alignment. These guidelines are here for reaching an appropriate level of alignment. Feel free to challenge the current state and to contribute!
+
+## Content
+
+* [Goals](#goals)
+* [One-by-one steps](#one-by-one-steps)
+* [Principles](#principles)
+* [Links](#links)
+
+## Goals
```
Comment (porcupine7) [inline GUIDELINES.md:12]: Nice overview!

File: GUIDELINES.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  * to sum up, there is a proper separation of concerns (separation of business rules, workflow and technical implementation) and the ability to reuse test code
+* **Simplify dependency tree creation** by using configuration and automatic dependency resolution
+* That **simplifies test setup** which makes it **cheaper to have bigger integration tests**
+* With that in mind it becomes worth to **invest more in integration tests** instead of unit tests
+  * which **reduces the use of mocks** and
+  * leads to **more realistic tests**
+
+## One-by-one steps
+
+* Create a module testing configuration in the test sources for each module in your project (e.g. for a module `app` create a file `AppModuleTestingConfiguration` with the following code: `val appModuleTestingConfiguration = moduleTestingConfiguration { ... }`; see below for more info on how to add dependencies to the configuration)
```
Comment (porcupine7) [inline GUIDELINES.md:27]: I think, it should state, that a testShared module is needed, where config und Steps reside.

File: GUIDELINES.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  * If modules depend on each other the module testing configuration should do the same, e.g. `val appModuleTestingConfiguration = moduleTestingConfiguration(dependentConfig1, dependentConfig2, ...) { ... }`, this automatically imports all dependent configurations
+* Create a test class for the _feature under test_ (e.g. `LoginTest`), so the test is agnostic of the underlying structures (classes, models, views, data structures, ...)
+  * Should it be that a test really just tests _one_ class, feel free to name it exactly like that (`LoginViewModelTest`)
+* Create a steps class for the feature under test (e.g. `LoginSteps`) which will contain all the `given`, `when` and `then` functions as well as dependency configuration and setup code
+  * The steps class should know about the setup of the test, so for example if the steps class tailored to test the integration of multiple classes, it should also be called something like `LoginIntegrationSteps` which tests a broad stack of classes (e.g. a `LoginViewModel` talking to an `AuthManager` and a `SessionStore`)
```
Comment (porcupine7) [inline GUIDELINES.md:35]: I don't think this naming convention is something which anybody does yet. Maybe get some feedback from more people here.

File: GUIDELINES.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  * Should it be that a steps class really just tests _one_ class, feel free to name it exactly like that (`LoginViewModelSteps`)
+  * In case the steps class' aim is purely to offer and act on a fake implementation of a class, name it with a `FakeSteps` suffix (e.g. `BackendGatewayFakeSteps` or `SessionStoreFakeSteps`)
+* Start with the test
```
Comment (porcupine7) [inline GUIDELINES.md:38]: I think this is up to individual preference, some might prefer to write steps first?

File: GUIDELINES.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  * In the test class you created add reference to the steps class with the `steps` function (e.g. `val sut by steps<LoginSteps>`) and call the value `sut` (system under test)
+  * Fill up the test with `@Test`-annotated test functions
+    * The test function can be named after a _scenario_ (e.g. `New user trying to log in`, `Weather condition is cloudy`)
+    * Only if necessary to distinguish test cases, also add more context like what's roughly expected (e.g. `New user trying to login - welcome to the app message shown`, `Existing user trying to login - welcome back message shown`)
```
Comment (porcupine7) [inline GUIDELINES.md:42]: Not sure, if I agree there. Having long test names is not a bad thing. Helps to see what failed immediately what went wrong in a report. And this will also be looked over in Audits. The more descriptive the better.

File: GUIDELINES.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    * Don't use names like `givenNewUser_whenLoggingIn_thenShowWelcomeToAppMessage` because
```
Comment (porcupine7) [inline GUIDELINES.md:43]: I agree, that it shouldn't be written in one long word. But written as sentence, this can imo be a valid test name

File: GUIDELINES.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+      * The test code in the test functions should already be easy enough to understand preconditions, actions and assumptions
+      * The more text you use the harder to distinguish test cases
+      * The less spaces you use the harder to read
+      * So please just use the test function names to make a meaningful distinctions to properly navigate between and understand test scenarios and cases
+  * For each test case, call functions that would reside in the steps class and prefix these functions with `given`, `when` or `then`
+    * Functions starting with `given` should set up the environment, as these define the assumptions that are made (e.g. `givenUser(email = "test@test.com", password = "supersecure")` indicating that the fake backend should act as if there was that user saved in the database). For the naming choose present tense (e.g. `givenUser` as a short form of "there is a user" or `givenInternetConnected`)
```
Comment (porcupine7) [inline GUIDELINES.md:49]: We talked about this already. Can you point me to some resources, which describe the given-when-then like this? I don't think it should be differentiated by tense. Given=pre-conditions / When=action which is test / Then = expected state after when. That's how understand this pattern, see https://archive.is/20140814232159/http://www.jroller.com/perryn/entry/given_when_then_and_how#selection-335.1-176.81

File: GUIDELINES.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    * Functions starting with `then` should perform assertions (e.g. `thenIsLoggedIn()`, `thenIsBusy()` or `thenUserIsCreated(user: User)`)
+    * By starting at the test to define which functions are needed in the steps you are more focused and get a faster overview of that's needed instead of getting into details too early
+  * Now the test should already be human-readable for non-tech people (with just the technical implementation in the steps class missing)
+  * Remember that in the test class all data structures and function calls should be domain-specific and therefore be modelled around real-world domain objects and data visible to a user, as far as possible (if the test subject is of total technical nature this doesn't apply of course)
+* Now think about which dependencies (classes or interfaces) will be under test and add new steps classes
+  * You can add a steps classs per class under test (e.g. `AuthManagerSteps`)
+  * As stated above (just to repeat)
+    * the steps classes' job is to know "how this class is to be tested"
```
Comment (porcupine7) [inline GUIDELINES.md:58]: I would say that a Steps class is basically the API to the production code. It should help to update the methods, signatures etc in production code without the need to update all the unit tests. This is often a complaint by people who don't write proper unit tests, and Steps classes help here :)

File: GUIDELINES.md (hunk 3 above)
Comment (porcupine7) [reply GUIDELINES.md:35]: In general I am not 100% sure if we should have multiple steps classes for the same class, just because it is real or mocked or whatever. As long as I can keep it real, I would do so. This is an advantage of sweetest we should also use. As stated in the beginning, we should prefer Steps classes which interact with real objects over faked ones, so I would just reuse them. And when it is really necessary, change the configuration in the steps class above. Not sure about it, but I would rather try to avoid having multiple steps classes and only do this when it is really necessary to have multiple ones.

File: GUIDELINES.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    * To conclude, a steps class usually can be considered responsible for inetracting with or mocking/faking the behavior of a class or interface under test
+  * You also need to think of dependencies
+    * A dependency in sweetest can be considered as an object that is needed by another or multiple other objects and usually contains logic
+    * Dependencies in sweetest are considered singletons
+    * You can tell sweetest which dependencies there are in the test system to let it automatically create and inject dependencies into constructors whenever certain dependencies are needed
+    * So when you request a dependency of a specific type it's automatically constructed and further dependencies in the constructor are automatically initialized as well (that relieves you of the burden of putting together huge test setups when having a lot of dependencies, which is even more cumbersome when writing integration tests)
+    * In essence, sweetest has an own small dependency injection framework included
+  * A dependency can be either mocked or real
+    * By default in sweetest all dependencies are considered mocks
```
Comment (porcupine7) [inline GUIDELINES.md:70]: Is this the case? But when I don't add it to the config, the tests don't run

File: GUIDELINES.md (hunk 1 above)
Comment (xa17d) [reply GUIDELINES.md:12]: Agree, this makes some design decisions clearer!

File: GUIDELINES.md [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+* [One-by-one steps](#one-by-one-steps)
+* [Principles](#principles)
+* [Links](#links)
+
+## Goals
+
+* Put a **layer of abstraction** on the system under test: we call these abstractions _steps_ (as derived from Cucumber)
+  * so if the system under test changes, the test system doesn't need to change
+  * so the test just tells _what_ is tested, not _how_ (tests become more business-centric, whenever possible, all the details are in the steps)
+  * so you can reuse test code
```
Comment (xa17d) [inline GUIDELINES.md:17]: Which one exactly? The _business-centric_ or the _technical implementation_ ones?

File: GUIDELINES.md [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+  * to sum up, there is a proper separation of concerns (separation of business rules, workflow and technical implementation) and the ability to reuse test code
+* **Simplify dependency tree creation** by using configuration and automatic dependency resolution
+* That **simplifies test setup** which makes it **cheaper to have bigger integration tests**
+* With that in mind it becomes worth to **invest more in integration tests** instead of unit tests
+  * which **reduces the use of mocks** and
+  * leads to **more realistic tests**
```
Comment (xa17d) [inline GUIDELINES.md:23]: But it also increases coupling between tests, right?
e.g. if I change one steps class, it could influence a lot of other step classes.

File: GUIDELINES.md [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+  * Always use the same package structure for test sources, e.g. if you app resides in a module `app` and has a base package `org.myorg.product.app` then put the module test configuration in exactly in the same package
+  * If there is a class `org.myorg.product.app.view.LoginViewModel` then the test class `LoginViewModelTest` and `LoginViewModelSteps` would reside in the exact same package, too
+  * If you have a dedicated module for test sources you should still stick to the same package structure, e.g. if the module is `:app:test` _don't_ add `test` to the package structure for test sources
+  * If modules depend on each other the module testing configuration should do the same, e.g. `val appModuleTestingConfiguration = moduleTestingConfiguration(dependentConfig1, dependentConfig2, ...) { ... }`, this automatically imports all dependent configurations
+* Create a test class for the _feature under test_ (e.g. `LoginTest`), so the test is agnostic of the underlying structures (classes, models, views, data structures, ...)
```
Comment (xa17d) [inline GUIDELINES.md:32]: I think this is important to underline: sweetest makes sense for a _feature under tests_. For some simple classes, plain old simple unit tests are the way to go (IMO).

File: GUIDELINES.md [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+  * Should it be that a test really just tests _one_ class, feel free to name it exactly like that (`LoginViewModelTest`)
+* Create a steps class for the feature under test (e.g. `LoginSteps`) which will contain all the `given`, `when` and `then` functions as well as dependency configuration and setup code
+  * The steps class should know about the setup of the test, so for example if the steps class tailored to test the integration of multiple classes, it should also be called something like `LoginIntegrationSteps` which tests a broad stack of classes (e.g. a `LoginViewModel` talking to an `AuthManager` and a `SessionStore`)
+  * Should it be that a steps class really just tests _one_ class, feel free to name it exactly like that (`LoginViewModelSteps`)
+  * In case the steps class' aim is purely to offer and act on a fake implementation of a class, name it with a `FakeSteps` suffix (e.g. `BackendGatewayFakeSteps` or `SessionStoreFakeSteps`)
```
Comment (xa17d) [inline GUIDELINES.md:37]: Hm, I think i most of the time called it `Mocked...Steps` or `Fake...Steps`, because it's more natural to read compared to a suffix

**Comment by eder7** [comment]
> I think it is important to get this right

agree 😄 

> The structure is a bit hard to follow, but as you said you want to change that anyway.

right, have a look on the WIP for the formatted version: https://github.com/mysugr/sweetest/blob/guidelines-formatted/GUIDELINES.md

> My suggestion is something like this:
> 
> 1. Simple tutorial
> 2. Explanation of sweetest inner workings
> 3. Naming conventions
> 4. Guidelines/Tips
> 
> That would make it imo a bit easier to read through. Because now there are many jumps (from concept level to technical details, naming conventions, tips etc)

that's right. as said it's WIP. a state to be challenged content-wise.
thanks for the input, i'll work towards that! hope it will be a good compromise!

> Also a Hello World project would be great!

i'm currently transferring the example in the project to our new standards. so it's in the works...

File: GUIDELINES.md [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+* Start with the test
+  * In the test class you created add reference to the steps class with the `steps` function (e.g. `val sut by steps<LoginSteps>`) and call the value `sut` (system under test)
```
Comment (xa17d) [inline GUIDELINES.md:39]: I personally don't like `sut`, but I guess that is a matter of taste.
Also, since you mentioned that steps are targetting bigger blocks than just a class (maybe even a whole feature), wouldn't it make sense to call the `sut` smiliar to the steps name?

e.g. I find this more readable:
```
backend.givenReturnsError()
login.whenLogin()
login.thenErrorIsShown()
```
compared to:
```
backend.givenReturnsError()
sut.whenLogin()
sut.thenErrorIsShown()
```

File: GUIDELINES.md (hunk 3 above)
Comment (eder7) [reply GUIDELINES.md:35]: i think that's already 3 questions you're addressing:

1. having separate steps classes for interacting with mock vs. a real implementation (i think it should be seperate)
2. having some kind of naming convention for steps classes (i think it makes sense, we should align on it though)
3. whether how much on an integration level we should test at all (i agree that integration level should be preferred, but that's anyways stated at many places, or at least it will)

File: GUIDELINES.md (hunk 4 above)
Comment (eder7) [reply GUIDELINES.md:38]: totally right! i'm a little split because i wanted to have a comprehensive walkthrough. and that implies some order. maybe i should take care of a formulation that leaves it more open to the dev...

File: GUIDELINES.md [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+  * Fill up the test with `@Test`-annotated test functions
+    * The test function can be named after a _scenario_ (e.g. `New user trying to log in`, `Weather condition is cloudy`)
```
Comment (xa17d) [inline GUIDELINES.md:41]: On the pump we tried to name test functions after requirements. So we don't name _scenarios_, but express what is expected to happen. And I think that would make sense here as well that would be e.g.
instead of `New user trying to log in` → `New user is asked for name, nationality and password`.
instead of `Existing user trying to log in` → `Existing user is only asked for password`.
instead of `Weather condition is cloudy` → `Background turns gray when weather condition is cloudy`.
What do you think?

File: GUIDELINES.md [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+    * Only if necessary to distinguish test cases, also add more context like what's roughly expected (e.g. `New user trying to login - welcome to the app message shown`, `Existing user trying to login - welcome back message shown`)
+    * Don't use names like `givenNewUser_whenLoggingIn_thenShowWelcomeToAppMessage` because
+      * The test code in the test functions should already be easy enough to understand preconditions, actions and assumptions
+      * The more text you use the harder to distinguish test cases
```
Comment (xa17d) [inline GUIDELINES.md:45]: Yes, it should be concise, but not I don't want to end up with test names like `new user`, `test login button`,...

File: GUIDELINES.md (hunk 7 above)
Comment (xa17d) [inline GUIDELINES.md:49]: I think `givenUser` is not very expressive. It should describe a state, e.g.
`backend.givenUserExists(...)`
`homescreen.givenUserIsLoggedIn(...)`

I remember having to look into implementation details of such functions, because I couldn't derive from the name what's actually happening. I know this is also a problem of simple tests becoming more complex and powerful.

File: GUIDELINES.md [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+    * Functions starting with `when` should perform actions, therefore they are named in progressive form (e.g. `whenGoingOffline`, `whenInitializing`, `whenLoggingInWith(email, password)`)
+    * Functions starting with `then` should perform assertions (e.g. `thenIsLoggedIn()`, `thenIsBusy()` or `thenUserIsCreated(user: User)`)
+    * By starting at the test to define which functions are needed in the steps you are more focused and get a faster overview of that's needed instead of getting into details too early
+  * Now the test should already be human-readable for non-tech people (with just the technical implementation in the steps class missing)
+  * Remember that in the test class all data structures and function calls should be domain-specific and therefore be modelled around real-world domain objects and data visible to a user, as far as possible (if the test subject is of total technical nature this doesn't apply of course)
```
Comment (xa17d) [inline GUIDELINES.md:54]: Reminds me that we should have more naming conventions like [Pump Hub Screens Naming Conventions](https://docs.google.com/presentation/d/1n_fKJGNAi2SJBSloKS3mgJG9oUowr__E0A8ct7vrgFg/edit#slide=id.p).

File: GUIDELINES.md [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+* Now think about which dependencies (classes or interfaces) will be under test and add new steps classes
+  * You can add a steps classs per class under test (e.g. `AuthManagerSteps`)
```
Comment (xa17d) [inline GUIDELINES.md:56]: Would you really create steps class per class or more per Sub-System?
Because just thinking, on a higher level (where the tests are written), I wouldn't care about an `AuthManager`, what I care about is e.g. does the user exist on the backend → `FakeBackendSteps`, etc.

File: GUIDELINES.md [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+  * As stated above (just to repeat)
+    * the steps classes' job is to know "how this class is to be tested"
+    * these will contain the `given`, `when` and `then` functions
+    * and if reasonable you can split steps classes by responsibility rather than classes, as in generally it's considered a better approach to test against concepts under test, not concrete classes; but in practice this turns out to be very hard to achieve. Testing against interfaces instead of concrete classes would be a good compromise between ease of test implementation and concern separation.
+    * If the steps class is purely here for faking or mocking the real behavior of a class you should name it accordingly (e.g. `AuthManagerMockSteps` or `BackendGatewayFakeSteps`)
```
Comment (xa17d) [inline GUIDELINES.md:61]: Ah, here you describe what I mentioned above 😅

File: GUIDELINES.md [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+    * To conclude, a steps class usually can be considered responsible for inetracting with or mocking/faking the behavior of a class or interface under test
```
Comment (xa17d) [inline GUIDELINES.md:62]:
```suggestion
    * To conclude, a steps class usually can be considered responsible for interacting with or mocking/faking the behavior of a class or interface under test
```

File: GUIDELINES.md (hunk 9 above)
Comment (xa17d) [reply GUIDELINES.md:70]: ...and I think this is the correct behavior. Automatically assume mock can lead to hard debugging sessions.

File: GUIDELINES.md (hunk 5 above)
Comment (eder7) [reply GUIDELINES.md:42]: i'm not sure what auditors exactly look at, but i think overall comprehensiveness is more important. in the end the test function name won't give anyone any kind of security or satisfying insight into the test. also the argument that you exactly know what failes when you look at the test name is not 100% correct. it can help narrowing down the issue but won't replace going deeply into the code in order to understand the problem. on top of that, the exact circumstances can way more precisely be extracted from the stack trace of the failed test. the abstracted style of tests like we do it in sweetest helps most in making the test more comprehensive.

so given the fact you usually spend most time reading and modifying tests my focus is more on the distinctiveness of the test function name over completeness. i think we already do it like that in our code, so it's probably not about our different styles but about the proper wording in these guidelines. we could use e.g. instead of "you should do this or that" something of the form like "you don't have to..." and moreover "make sure it's descriptive enough to understand what's tested as fast as possible, meanwhile making it distinctive in order to be able to havigate through the test fast" and add some examples...

File: GUIDELINES.md [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+* How to add a dependency
+  * You need to put each dependency you use in sweetest into the **module testing configuration** (see above)
+    * Add a dependency by adding `dependency any of<LoginViewModel>` to the configuration (you need to do this before you introduce the dependency to any test or steps class!)
+    * Consider mock vs. real
+      * For example, if you are testing just the `LoginViewModel`, you set sweetest up so it creates a _real_ `LoginViewModel` only (if you don't tell sweetest it will always create a mock for that type)
+      * For example, if you are testing the integration of the `LoginViewModel` with `AuthManager` and `SessionStore` then you need to tell sweetest to create real instances for all of these (`BackendGateway` will remain mocked, though)
+    * Set mode constraints in configuration
+      * In your test systems, if you want to have a certain type always to be **real**, then you can tell sweetest in the configuration like this: `dependency realOnly of<LoginViewModel>` (that tells that `LoginViewModel` might not be used as a dependency in a mocked form; for `LoginViewModel` that makes total sense)
```
Comment (xa17d) [inline GUIDELINES.md:87]: Can we please get rid of the `realOnly` or `mockOnly` configs?
I often had the problem that somebody assumed "that always has to be mock" or "that's always real" and I needed the exact opposite. It's fix such a early mis-decision, it involves changing loads of files...

File: GUIDELINES.md (hunk 6 above)
Comment (eder7) [reply GUIDELINES.md:43]: i think it's a redundancy to repeat yourself in the name of the test function _and_ in the test body. when the test is properly abstracted (preferrably around business domain concepts) you'll find exactly these functions in the body:

```
givenNewUser()
whenLoggingIn()
thenShowWelcomeToAppMessage()
```

as the test body is so easy to read i think it becomes obsolete to be that precise in the function name. see also arguments in the comment above...

File: GUIDELINES.md (hunk 7 above)
Comment (eder7) [reply GUIDELINES.md:49]: there are a lot of different opinions on that, e.g. in [fowler's article](https://martinfowler.com/bliki/GivenWhenThen.html) you linked once the "i" form is used. this is just a proposal, i'm for streamlining this. on jroller.com also present tense is used for "given" statements and actually i see nothing noteworthy that conflicts with his proposal. what do you prefer?

File: GUIDELINES.md (hunk 10 above)
Comment (eder7) [reply GUIDELINES.md:17]: the only thing you can reuse is the steps class, so the technical implementation. should i outline that clearer?

File: GUIDELINES.md (hunk 11 above)
Comment (eder7) [reply GUIDELINES.md:23]: that's right, so there's pros and cons. it becomes just more important to do abstraction right. do you see a big impendiment?

File: GUIDELINES.md (hunk 13 above)
Comment (eder7) [reply GUIDELINES.md:37]: hm, and what about the sorting? steps classes could be harder to spot then...

File: GUIDELINES.md (hunk 14 above)
Comment (eder7) [reply GUIDELINES.md:39]: i think `sut` makes only sense when there is just _one_ steps class in the test. As an example have a look at the `GetZoneOffsetChangesFromTimedEntriesUseCaseTest` and we use it a lot in our tests (>200 occurrences). but otherwise i agree with you that the names should be distinctive.

File: GUIDELINES.md (hunk 15 above)
Comment (eder7) [reply GUIDELINES.md:41]: i like the proposals, but would put it as "if it makes sense", especially to distinguish test function from each other. i want to add them as better examples, thanks!
i'll make a differentiated proposal also depending on the outcome of [that discussion](https://github.com/mysugr/sweetest/pull/24#discussion_r401611362). could you please add your opinion there?

File: GUIDELINES.md (hunk 16 above)
Comment (eder7) [reply GUIDELINES.md:45]: totally valid, they should have a minimum information value!

File: GUIDELINES.md (hunk 9 above)
Comment (eder7) [reply GUIDELINES.md:70]: what i wanted to say here is, if something isn't specified as real it's automatically a mock, will update the wording...

**Comment by eder7** [comment]
> This describes status quo, right? Not how it'll evolve.

it describes partly what we already do and partly what's not yet conventionalized. so actually both @xa17d

File: GUIDELINES.md (hunk 12 above)
Comment (eder7) [reply GUIDELINES.md:32]: i added `Don't use sweetest if it makes no sense (e.g. classes which will never be tested in an integrated fashion)` and `It is not the goal to overcomplicate tests`

File: GUIDELINES.md (hunk 16 above)
Comment (eder7) [reply GUIDELINES.md:45]: i tried to find a more balanced wording, how do you like it now?

File: GUIDELINES.md (hunk 18 above)
Comment (eder7) [reply GUIDELINES.md:56]: i think you're right. as we consider steps classes as a way of abstracting technical implementation these should probably be more independent of the _concrete_ implementation and structure. i think it depends on the case. by having steps classes like `LoginViewModelIntegrationSteps` we integrate the whole system anyways already. changes in the structure should not affect the test class, only the steps class, in an optimal case. so according to your input i'd call it `LoginIntegrationSteps` or simply `LoginSteps`?

File: GUIDELINES.md (hunk 21 above)
Comment (eder7) [reply GUIDELINES.md:87]: i was working on getting rid of the whole module configuration as such, couldn't find the priority to actually finish it. in the meantime we could make all dependencies `any` in the config and push all further configuration down to the tests and steps by convention (here). what do you think of that?

File: GUIDELINES.md (hunk 4 above)
Comment (eder7) [reply GUIDELINES.md:39]: will take that into account in the formatted version...

File: GUIDELINES.md (hunk 15 above)
Comment (porcupine7) [reply GUIDELINES.md:42]: I would put the naming conventions in an own section and then structure it similar to our [Variable guidelines](https://github.com/mysugr/shared-android/blob/develop/documentation/naming-conventions/variables.md). Otherwise it is a bit all over the place in the document. So you can add wrong, ok and good examples. I agree with Daniel that test names should at least define what is expected to happen.

File: GUIDELINES.md (hunk 18 above)
Comment (eder7) [reply GUIDELINES.md:57]: proposal:

There are multiple options how steps classes can be named; some examples:

1. As abstraction of a class (`LoginViewModelSteps`): use this if a steps class solely concentrates on interacting with or mocking/faking a single specific class or interface 
2. As abstraction of an integration of classes (`LoginViewModelIntegrationSteps`): in cases where the class or interface is tested in integration with other classes
3. As abstraction of a feature or subsystem (`LoginSteps`): in cases where

File: GUIDELINES.md (hunk 5 above)
Comment (porcupine7) [reply GUIDELINES.md:43]: > the argument that you exactly know what failes when you look at the test name is not 100% correct. 

That depends how the tests are written, I like this [article](https://proandroiddev.com/seven-principles-of-great-unit-tests-adapted-for-android-342515f98ef2) a lot. The section 2. Independent describe it well I think.

As said, examples would help to clarify the naming conventions.

File: GUIDELINES.md (hunk 7 above)
Comment (porcupine7) [reply GUIDELINES.md]: I think given should describe pre-conditions of the test (which can be actions too). When = actions of the scenario. And then = the expected state afterwards. So, when we consider given defines the pre-conditions for the test, it can very well be also in the past tense. E.g. givenNeverLoggedIn could be also a precondition. 

I just don't think enforcing a tense should be in the guidelines. We can of course start to do it, but the more important part is that these keywords are used in the proper context of the test. An example would help.

File: GUIDELINES.md [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+  * Should it be that a steps class really just tests _one_ class, feel free to name it exactly like that (`LoginViewModelSteps`)
+  * In case the steps class' aim is purely to offer and act on a fake implementation of a class, name it with a `FakeSteps` suffix (e.g. `BackendGatewayFakeSteps` or `SessionStoreFakeSteps`)
+* Start with the test
+  * In the test class you created add reference to the steps class with the `steps` function (e.g. `val sut by steps<LoginSteps>`) and call the value `sut` (system under test)
+  * Fill up the test with `@Test`-annotated test functions
+    * The test function can be named after a _scenario_ (e.g. `New user trying to log in`, `Weather condition is cloudy`)
+    * Only if necessary to distinguish test cases, also add more context like what's roughly expected (e.g. `New user trying to login - welcome to the app message shown`, `Existing user trying to login - welcome back message shown`)
+    * Think twice if you see a test named `givenNewUser_whenLoggingIn_thenShowWelcomeToAppMessage`
```
Comment (floschu) [inline GUIDELINES.md:44]:
```suggestion
    * Think twice before naming a test `givenNewUser_whenLoggingIn_thenShowWelcomeToAppMessage`
```

File: GUIDELINES.md (hunk 18 above)
Comment (floschu) [reply GUIDELINES.md:57]: Maybe add a naming conventions table at the bottom, as mentioned above already.

File: GUIDELINES.md [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+    * As you can see, you can control default configuration and initialization behavior on a module basis _and_ on a steps basis
+  * The easiest way to approach configuration is to just allow what's necessary on a module level and make it more specific on the test or steps level
+    * E.g. if there is a `AuthManager` which can be used as a mock and a real instance as well, configure it as `any` in module configuration
+    * If in a test system there is an integration setup which requires `AuthManager` to be mock and a different setup which required `AuthManager` to be a real instance, consider the following:
+    * Create a steps class which is specialized on mocking `AuthManager` and call it `AuthManagerMockSteps`
+      * there you also configure the `AuthManager` dependency to be a mock only by using `requireMockOnly<>`
+* If you have done that we should already have a structure like this
+  * If possible, the test class refers to a **user-facing feature** under test (e.g. "LoginTest", "TemperatureDisplayTest", "DataLoggingTest", ...)
+  * The the test functions are named after scenarios
```
Comment (floschu) [inline GUIDELINES.md:129]: All of the above is very detailed 👍 
However I would much rather have some short examples at the top for a quick overview.
As someone who codes but has never seen this library in action before or not worked with it, I think such an example could give a very quick grasp:

``` kotlin
/**
 * Explanation
 */
val appModuleTestingConfiguration = moduleTestingConfiguration(
    dependentConfig1 // explanation
) {
    dependency any of<Whatever>()  // explanation
    dependency mockOnly of<Whatever2>()  // explanation
    ....
}

// needs to be tested
class Someclass @Inject constructor(val whatever: Whatever) { ... }

/**
 * Explanation
 */
class SomeClassTest : BaseJUnitTest(appModuleTestingConfiguration) {
    private val sut by steps<SomeClassSteps>()

    @Test
    fun `bla blue bli`() = sut {
        given()
        when()
        then()
    }
}

/**
 * Explanation
 */
class SomeClassSteps : (
    testContext: TestContext // explanation
) : BaseSteps(testContext, appModuleTestingConfiguration) {
    ....
}
```

File: GUIDELINES.md (hunk 5 above)
Comment (eder7) [reply GUIDELINES.md:43]: the respective paragraph in this article is not 100% conclusive for me as it doesn't give me an exact reason, but i think this is probably too much into detail for this guidelines. i'm considering removing that point...

File: GUIDELINES.md (hunk 6 above)
Comment (eder7) [reply GUIDELINES.md]: maybe a hint that the abstracted style inside the test functions helps creating more context for the test to analyze is enough here then instead of discouraging the use of long function names.

File: GUIDELINES.md (hunk 7 above)
Comment (eder7) [reply GUIDELINES.md]: i think also this is a candidate for _not_ having it in the guidelines. in the end the guidelines should also be quick to understand, so it helps to not overcomplicate things.

File: GUIDELINES.md (hunk 4 above)
Comment (eder7) [reply GUIDELINES.md:39]: hm, now that i'm in the text again i realize why it makes sense to create the skeleton of the steps class _before_: you can integrate it as a `sut` member in the test, add calls to `sut` in test functions and let the IDE's refactoring functions auto-create these functions. if the steps class weren't yet here the calls would go nowhere, and the auto creation functions would not work. i'll describe it like that.
